File: leetcode/java/longestcommonprefix.java
/*
 * find the longest common prefix string amongst an array of strings
 * */
import java.io.*;
import java.util.*;

public class longestcommonprefix{
    public String longestCommonPrefix(String[] strs){
        int n = strs.length;
        if(n==0)    return new String();
        if(n==1)    return new String(strs[0]);
        int length = strs[0].length();
        for(int i=1;i<n;i++){
            if(length > strs[i].length()){
                length = strs[i].length();
            }
        }
        StringBuilder builder = new StringBuilder();
        int i=0;
        for(;i<length;++i){
            Character ch = new Character(strs[0].charAt(i));
            int j;
            for(j=1;j<n && ch.equals(strs[j].charAt(i));++j);
            if(j<n)    break;
            builder.append(ch);
        }
        return builder.toString();
    }

    public static void main(String[] args){
        return;
    }
}

File: leetcode/src/longestCommonPrefix.cpp
/*
 * find the longest common prefix string among an array of strings
 * */
#include <string>
#include <cstring>
#include <vector>
#include <algorithm>

using namespace std;

class Solution{

public:
    /*
     * time O(n^2)
     * */
    string longestCommonPrefix(const vector<string>& strs){
        if(strs.empty())    return string();

        const int n = strs.size();
        if(n == 1)    return string(strs[0]);
        int length = strs[0].size();

        for(int i = 1; i < n; i++){
            length = min(length, (int)strs[i].size());
        }

        char res[length + 1];
        memset(res, 0, sizeof(res));

        int i = 0;
        for(; i < length; ++i){
            int j = 1;
            for(; j < n && strs[j][i] == strs[j-1][i]; ++j);
            if(j < n)    break;

            res[i] = strs[0][i];
        }
        res[i] = '\0';

        return string(res);
    }
};
